2-WebApp-graph-user/2-5-HybridFlow/Services/ConfidentialClientApplicationService.cs

```diff
@@ -0,0 +1,46 @@
+using System.Security.Claims;
+using Microsoft.Identity.Client;
+using Microsoft.Identity.Web;
+
+namespace WebApp_OpenIDConnect_DotNet.Services
+{
+    public class ConfidentialClientApplicationService : IConfidentialClientApplicationService
+    {
+        private static IConfidentialClientApplication? _confidentialClientApplication;
+        private static IConfidentialClientApplication ConfidentialClientApplication
+        {
+            get
+            {
+                if (_confidentialClientApplication == null)
+                {
+                    var config = AuthenticationConfig.ReadFromJsonFile("appsettings.json");
+                    _confidentialClientApplication = ConfidentialClientApplicationBuilder.Create(config.ClientId)
+                        .WithClientSecret(config.ClientSecret)
+                        .WithRedirectUri(config.RedirectUri)
+                        .WithAuthority(new Uri(config.Authority))
+                        .Build();
+                    
+                    _confidentialClientApplication.AddInMemoryTokenCache();
+                }
+
+                return _confidentialClientApplication;
+            }
+        }
+       public async Task<AuthenticationResult> GetAuthenticationResultAsync(string[] scopes, string code, string codeVerifier)
+        {
+            return await ConfidentialClientApplication
+                .AcquireTokenByAuthorizationCode(scopes, code)
+                .WithPkceCodeVerifier(codeVerifier)
+                .WithSpaAuthorizationCode(true)
+                .ExecuteAsync();
+        }
+
+        public async Task RemoveAccount(string identifier) {
+            var userAccount = await ConfidentialClientApplication.GetAccountAsync(identifier);
+            if (userAccount != null)
+            {
+                await ConfidentialClientApplication.RemoveAsync(userAccount);
+            }
+        }
+    }
+}
```

Comment:
This deviates from the pattern I've seen in other samples. Instead of having one off instances of the `ConfidentialClientApplication` it's isolated within a service that can be used via dependency injection for .NET Core style apps.

The idea here is that it would make the methods easier to access in other parts of the application (such as controllers) and abstracting away some of the complexity. E.g. making a single method call for acquiring an `AuthenticationResult` object from the authentication flow as `GetAuthenticationResultAsync`.

The **WithPkceCodeVerifier** method is required to have PKCE work properly with this sample.

This also makes unit testing and mocking easier.